scapy/contrib/scada/iec104/iec104_fields.py

```diff
@@ -0,0 +1,144 @@
+# This file is part of Scapy
+# See http://www.secdev.org/projects/scapy for more information
+# Copyright (C) Thomas Tannhaeuser <hecke@naberius.de>
+# This program is published under a GPLv2 license
+
+# scapy.contrib.description = IEC-60870-5-104 layer specific fields
+# scapy.contrib.status = loads
+
+"""
+    field type definitions used by iec 60870-5-104 layer (iec104)
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+    :description:
+
+        This file provides field definitions used by the IEC-60870-5-104
+        implementation. Some of those fields are used exclusively by iec104
+        (e.g. IEC104SequenceNumber) while others (LESignedShortField) are
+        more common an may be moved to fields.py.
+
+        normative references:
+            - EN 60870-5-104:2006
+            - EN 60870-5-4:1993
+            - EN 60870-5-4:1994
+"""
+import struct
+
+from scapy.compat import orb
+from scapy.fields import Field, ThreeBytesField, BitField
+from scapy.volatile import RandSShort
+
+
+class LESignedShortField(Field):
+    """
+    little endian signed short field
+    """
+    def __init__(self, name, default):
+        Field.__init__(self, name, default, "<h")
+
+
+class IEC60870_5_4_NormalizedFixPoint(LESignedShortField):
+    """
+    defined as typ 4.1 in EN 60870-5-4:1993, sec. 5.4.1 (p. 10)
+    """
+
+    def i2repr(self, pkt, x):
+        """
+        show the fixed fp-number and its signed short representation
+        """
+
+        return '{} ({})'.format(self.i2h(pkt, x), x)
+
+    def i2h(self, pkt, x):
+        return x / 32768.
+
+    def randval(self):
+        # ToDo: this could also be implemented by adding fmt h+RandSShort to
+        # randval@class Field - should we ?!?
+        return RandSShort()
+
+
+class LEIEEEFloatField(Field):
+    """
+    little endian IEEE float field
+    """
+    def __init__(self, name, default):
+        Field.__init__(self, name, default, "<f")
+
+
+class LEThreeBytesField(ThreeBytesField):
+    """
+    little endian three bytes field
+    """
+    def __init__(self, name, default):
+        ThreeBytesField.__init__(self, name, default)
+
+    def addfield(self, pkt, s, val):
+        data = struct.pack(self.fmt, self.i2m(pkt, val))[1:4][::-1]
+        return s + data
+
+    def getfield(self, pkt, s):
+        data = s[:3][::-1]
+        return s[3:], self.m2i(pkt, struct.unpack(self.fmt, b"\x00" + data)[0])
+
+
+class IEC104SequenceNumber(Field):
```

Comment:
Looks very much like https://github.com/secdev/scapy/pull/1924

You think you could use it if it was merged ?

Reply:
If I understand the description of #1924 right, this field consumes bytes as long as the LSB is one and stops if the LSB is zero. I don't know how to use this field here as the LSB 'within' the field can take any value (1/0) as all bits are taken (not only the 7 upper ones). 

But if I got something wrong (likely) - I happily switch to some standard type as I always fear that my layer specific types contain some monsters hiding in the dark...

Reply:
It says

> `extension_bit is the bit number [0-7] of the extension bit in the byte`

Reply:
but... as far as I could see - the BitExtendedField works byte wise - but IEC104SequenceNumber spans two bytes - so the LSB of the 'higher byte' may or may not contain a logical one. 

There is a already a test (https://github.com/speakinghedge/scapy/blob/iec-60870-5-104/scapy/contrib/iec104.uts#L44) that contains some test data that may outline the issue:

```
test_data = {
    1: b'\x02\x00',
    2: b'\x04\x00',
    14 : b'\x1c\x00',
    16 : b'\x20\x00',
    73 : b'\x92\x00',
    127: b'\xfe\x00',
    128: b'\x00\x01',
    129: b'\x02\x01',
    253: b'\xfa\x01',
    254: b'\xfc\x01',
    255: b'\xfe\x01',
    5912: b'\x30\x2e',
    31282: b'\x64\xf4',
    32767: b'\xfe\xff'
}
```
Only the LSB of the first byte is always zero.
The second one may or may not be zero - independently of the remaining bits of this byte - so ExtendedField won't work here (again: as far as I understand. maybe I got something fundamental wrong - or your request implies "extend the BitExtendedField to allow extension_bit > 7...")

I still don't get it.